chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+
+
+class SdistMetadataFetcher(object):
+    """This is the "correct" way to get name and version from an sdist."""
+    # https://git.io/vQkwV
+    _SETUPTOOLS_SHIM = (
+        "import setuptools, tokenize;__file__=%r;"
+        "f=getattr(tokenize, 'open', open)(__file__);"
+        "code=f.read().replace('\\r\\n', '\\n');"
+        "f.close();"
+        "exec(compile(code, __file__, 'exec'))"
+    )
+
+    def __init__(self, osutils=None):
+        # type: (OSUtils) -> None
+        if osutils is None:
+            osutils = OSUtils()
+        self._osutils = osutils
+
+    def _parse_pkg_info_file(self, filepath):
+        # type: (str) -> Message
+        with open(filepath, 'r') as f:
+            data = f.read()
+        parser = FeedParser()
+        parser.feed(data)
+        return parser.close()
+
+    def _generate_egg_info(self, package_dir):
+        # type: (str) -> str
+        setup_py = self._osutils.joinpath(package_dir, 'setup.py')
+        script = self._SETUPTOOLS_SHIM % setup_py
+
+        cmd = [sys.executable, '-c', script, '--no-user-cfg', 'egg_info']
+        egg_info_dir = self._osutils.joinpath(package_dir, 'egg-info')
+        self._osutils.makedirs(egg_info_dir)
+        cmd += ['--egg-base', 'egg-info']
+        p = subprocess.Popen(cmd, cwd=package_dir,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        p.communicate()
+        self._osutils.joinpath(egg_info_dir)
+        info_contents = self._osutils.get_directory_contents(egg_info_dir)
+        assert len(info_contents) == 1
+        pkg_info_path = self._osutils.joinpath(
+            egg_info_dir, info_contents[0], 'PKG-INFO')
+        return pkg_info_path
+
+    def _unpack_sdist_into_dir(self, sdist_path, unpack_dir):
+        # type: (str, str) -> str
+        if sdist_path.endswith('zip'):
+            with zipfile.ZipFile(sdist_path, 'r') as z:
+                z.extractall(unpack_dir)
+        elif sdist_path.endswith('.tar.gz'):
+            with tarfile.open(sdist_path, 'r:gz') as tar:
+                tar.extractall(unpack_dir)
+        else:
+            raise InvalidSourceDistributionNameError(sdist_path)
+        # There should only be one directory unpacked.
+        contents = self._osutils.get_directory_contents(unpack_dir)
+        assert len(contents) == 1
+        return self._osutils.joinpath(unpack_dir, contents[0])
+
+    def get_package_name_and_version(self, directory, filename):
+        # type: (str, str) -> Tuple[str, str]
+        sdist_path = self._osutils.joinpath(directory, filename)
+        with self._osutils.tempdir() as tempdir:
+            package_dir = self._unpack_sdist_into_dir(sdist_path, tempdir)
+            pkg_info_filepath = self._generate_egg_info(package_dir)
+            metadata = self._parse_pkg_info_file(pkg_info_filepath)
+            name = metadata['Name']
+            version = metadata['Version']
+        return name, version
+
+
+class PipWrapper(object):
```

Comment:
I subclassed it for testing purposes as well and wanted to ensure that the type signature contract is not broken by the testing class. This way mypy will catch it if either of them get out of sync. I could also move the logic from the `SubprocessPip` up to `PipWrapper`and just have the testing inherit from `PipWrapper` and override the `main` method. I thought it was clearer what the purpose of everything was this way.